Spigot-Server-Patches/0424-Seed-based-feature-search.patch

```diff
@@ -0,0 +1,110 @@
+     public int e() {
+         return this.z << 4;
+     }
+diff --git a/src/main/java/net/minecraft/server/StructureGenerator.java b/src/main/java/net/minecraft/server/StructureGenerator.java
+index e8ce2ecf2..acfe732af 100644
+--- a/src/main/java/net/minecraft/server/StructureGenerator.java
++++ b/src/main/java/net/minecraft/server/StructureGenerator.java
+@@ -109,6 +109,15 @@ public abstract class StructureGenerator<C extends WorldGenFeatureConfiguration>
+                             if (flag1 || flag2) {
+                                 ChunkCoordIntPair chunkcoordintpair = this.a(chunkgenerator, seededrandom, j, k, i1, j1);
+                                 if (!world.getWorldBorder().isChunkInBounds(chunkcoordintpair.x, chunkcoordintpair.z)) { continue; } // Paper
++                                // Paper start - seed based feature search
++                                if (world.paperConfig.seedBasedFeatureSearch) {
++                                    BiomeManager biomeManager = world.getBiomeManager().withProvider(chunkgenerator.getWorldChunkManager());
++                                    BiomeBase biomeBase = biomeManager.getBiome(new BlockPosition(chunkcoordintpair.getBlockX() + 9, 0, chunkcoordintpair.getBlockZ() + 9));
++                                    if (!shouldGenerate(biomeManager, chunkgenerator, seededrandom, chunkcoordintpair.x, chunkcoordintpair.z, biomeBase)) {
++                                        continue;
++                                    }
++                                }
++                                // Paper end
+                                 StructureStart structurestart = world.getChunkAt(chunkcoordintpair.x, chunkcoordintpair.z, ChunkStatus.STRUCTURE_STARTS).a(this.b());
+ 
+                                 if (structurestart != null && structurestart.e()) {
+@@ -165,6 +174,7 @@ public abstract class StructureGenerator<C extends WorldGenFeatureConfiguration>
+         return new ChunkCoordIntPair(i + k, j + l);
+     }
+ 
++    public boolean shouldGenerate(BiomeManager biomemanager, ChunkGenerator<?> chunkgenerator, Random random, int chunkX, int chunkZ, BiomeBase biomebase) { return a(biomemanager, chunkgenerator, random, chunkX, chunkZ, biomebase); } // Paper - OBFHELPER
+     public abstract boolean a(BiomeManager biomemanager, ChunkGenerator<?> chunkgenerator, Random random, int i, int j, BiomeBase biomebase);
+ 
+     public abstract StructureGenerator.a a();
+diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
+index 5460ace8f..999338cfc 100644
+--- a/src/main/java/net/minecraft/server/World.java
++++ b/src/main/java/net/minecraft/server/World.java
+@@ -1584,6 +1584,7 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
+         return this.methodProfiler;
+     }
+ 
++    public BiomeManager getBiomeManager() { return d(); } // Paper - OBFHELPER
+     @Override
```

Comment:
Override should be moved to next line for effective use of the OPFHELPER.

Reply:
I'm unsure what you mean by that, the `@Override` is for the d() method which overrides the one in the `IWorldReader` interface. I guess you mean that the OBFHELPER should be in the interface via a default method or something like that?